tools/tpm2_ptool.py

```diff
@@ -1233,6 +1283,206 @@ def __call__(self, args):
 
                 print('Added key as label: "{keylabel}"'.format(keylabel=keylabel))
 
+@commandlet("importrsa")
+class ImportRsaCommand(Command):
+    '''
+    Imports a rsa key to a token within a tpm2-pkcs11 store.
+    '''
+
+    # adhere to an interface
+    # pylint: disable=no-self-use
+    def generate_options(self, group_parser):
+        group_parser.add_argument(
+            '--privkey',
+            help='Full path of the private key to be imported.\n',
+            required=True)
+        group_parser.add_argument(
+            '--label',
+            help='The tokens label to add a key too.\n',
+            required=True)
+        group_parser.add_argument(
+            '--key-label',
+            help='The label of the key imported.\n')
+        group_parser.add_argument(
+            '--algorithm',
+            help='The type of the key.\n',
+            choices=[ 'rsa', 'rsa1024', 'rsa2048', 'aes128', 'aes256' ],
+            required=True)
+        group_parser.add_argument(
+            '--id',
+            help='The key id. Defaults to a random 8 bytes of hex.\n',
+            default=binascii.hexlify(os.urandom(8)))
+        pinopts = group_parser.add_mutually_exclusive_group(required=True)
+        pinopts.add_argument(
+            '--sopin',
+            help='The Administrator pin,'),
+        pinopts.add_argument(
+            '--userpin',
+            help='The User pin.\n'),
+
+    def __call__(self, args):
+
+
+        path = args['path']
+        privkey = args['privkey']
+        label = args['label']
+        keylabel= args['key_label']
+
+        id = args['id']
+
+        with Db(path) as db:
+
+            token = db.gettoken(label)
+
+            token_config =  dict_from_kvp(token['config'])
+
+            pobj = db.getprimary(token['pid'])
+
+            with TemporaryDirectory() as d:
+                tpm2 = Tpm2(d, path)
+
+                # Get the primary object encrypted auth value and sokey information
+                # to decode it. Based on the incoming pin
+                is_so = args['sopin'] != None
+                if is_so:
+                    pin = args['sopin']
+                    pinpobjauthkeysalt = token['sopobjauthkeysalt']
+                    pinpobjauthkeyiters = token['sopobjauthkeyiters']
+                    pinpobjauth = token['sopobjauth']
+                else:
+                    pin = args['userpin']
+                    pinpobjauthkeysalt = token['userpobjauthkeysalt']
+                    pinpobjauthkeyiters = token['userpobjauthkeyiters']
+                    pinpobjauth = token['userpobjauth']
```

Comment:
This duplicates addkey command a lot. Refactor this so common blocks are moved and shared between the two. Perhaps use class inheritance and provide just the parts that differ....

I think you would be able to do a super class called generickey or something like that, and then have each addkey and importkey commandlets inherit that generickey class. The call could call the super class and then add the missing functionality via some interface.

They both do the same thing except one does inport and the other calls create. See how I did *AddEmptyTokenCommand* command.

Reply:
Definitely, just wanted to check if this worked given the urgency.